src/main/java/org/squiddev/cobalt/function/LuaInterpreter.java

```diff
@@ -640,6 +655,850 @@ A C R(A+3), ... ,R(A+2+C):= R(A)(R(A+1),
+				}
+
+				if (c > 0xff) {
+					final LuaValue konstRight = k[c & 0x0ff];
+					return cont.apply(di -> di.stack[a] = OperationHelper.mod(state, di.stack[b], konstRight, b, c));
+				}
+
+				return cont.apply(di -> di.stack[a] = OperationHelper.mod(state, di.stack[b], di.stack[c], b, c));
+			}
+
+			case OP_POW: { // A B C: R(A):= RK(B) ^ RK(C)
+				final int b = (i >>> POS_B) & MAXARG_B;
+				final int c = (i >> POS_C) & MAXARG_C;
+				if (b > 0xff) {
+					final LuaValue konstLeft = k[b & 0x0ff];
+					if (c > 0xff) {
+						final LuaValue konstRight = k[c & 0x0ff];
+						return cont.apply(di -> di.stack[a] = OperationHelper.pow(state, konstLeft, konstRight, b, c));
+					}
+
+					return cont.apply(di -> di.stack[a] = OperationHelper.pow(state, konstLeft, di.stack[c], b, c));
+				}
+
+				if (c > 0xff) {
+					final LuaValue konstRight = k[c & 0x0ff];
+					return cont.apply(di -> di.stack[a] = OperationHelper.pow(state, di.stack[b], konstRight, b, c));
+				}
+
+				return cont.apply(di -> di.stack[a] = OperationHelper.pow(state, di.stack[b], di.stack[c], b, c));
+			}
+
+			case OP_UNM: { // A B: R(A):= -R(B)
+				final int b = (i >>> POS_B) & MAXARG_B;
+
+				if (b > 0xff) {
+					final LuaValue konst = k[b & 0x0ff];
+					return cont.apply(di -> di.stack[a] = OperationHelper.neg(state, konst, b));
+				}
+
+				return cont.apply(di -> di.stack[a] = OperationHelper.neg(state, di.stack[b], b));
+			}
+
+			case OP_NOT: { // A B: R(A):= not R(B)
+				final int b = (i >>> POS_B) & MAXARG_B;
+				return cont.apply(di -> di.stack[a] = di.stack[b].toBoolean() ? FALSE : TRUE);
+			}
+
+			case OP_LEN: { // A B: R(A):= length of R(B)
+				final int b = (i >>> POS_B) & MAXARG_B;
+				return cont.apply(di -> di.stack[a] = OperationHelper.length(state, di.stack[b], b));
+			}
+
+			case OP_CONCAT: { // A B C: R(A):= R(B).. ... ..R(C)
+				final int b = (i >>> POS_B) & MAXARG_B;
+				final int c = (i >> POS_C) & MAXARG_C;
+				final int total = c - b + 1;
+
+				return cont.apply(di -> {
+					di.top = c + 1;
+					concat(state, di, di.stack, di.top, total);
+					di.stack[a] = di.stack[b];
+					di.top = b;
+				});
+			}
+
+			case OP_JMP: { // sBx: pc+=sBx
+				final int offset = ((i >>> POS_Bx) & MAXARG_Bx) - MAXARG_sBx + 1;
+				return raw.apply(di -> new EvalCont(pc + offset));
+			}
+
+			// TODO precompute comparisons of constants
+			case OP_EQ: { // A B C: if ((RK(B) == RK(C)) ~= A) then pc++
+				final int b = (i >>> POS_B) & MAXARG_B;
+				final int c = (i >> POS_C) & MAXARG_C;
+				final boolean aNotZero = a != 0;
+
+				// We assume the next instruction is a jump and read the branch from there.
+				final int offset = ((code[pc + 1] >> POS_Bx) & MAXARG_Bx) - MAXARG_sBx + 2;
+
+				if (b > 0xff) {
+					final LuaValue konstLeft = k[b & 0x0ff];
+					if (c > 0xff) {
+						final LuaValue konstRight = k[c & 0x0ff];
+						return raw.apply(di -> {
+							if (OperationHelper.eq(state, konstLeft, konstRight) == aNotZero) {
```

Comment:
Ternary expressions? Heck, could even do `new EvalCont(_ ? _ : _)`.

Reply:
The array access in `k[c & 0x0ff]` is generally unsafe, so we need a block around the lambda.

Reply:
I meant within the lambda body itself. Or am I misunderstanding?

Reply:
My bad, I thought you were referring to the surrounding `if`. Good point!